plugins/image-source/obs-slideshow-mk2.c

```diff
@@ -0,0 +1,1192 @@
+				  bool next)
+{
+	if (ssd->randomize) {
+		return random_file(ssd, slide_idx);
+	} else if (next) {
+		return (++slide_idx >= ssd->files.num) ? 0 : slide_idx;
+	} else {
+		return (slide_idx == 0) ? ssd->files.num - 1 : slide_idx - 1;
+	}
+}
+
+/* ------------------------------------------------------------------------- */
+
+static const char *ss_getname(void *unused)
+{
+	UNUSED_PARAMETER(unused);
+	return obs_module_text("SlideShow");
+}
+
+/* adds a new file to the files list. used only in update */
+static inline void add_file(image_file_array_t *new_files, const char *path)
+{
+	struct image_file_data data;
+	data.path = bstrdup(path);
+	da_push_back(*new_files, &data);
+}
+
+extern bool valid_extension(const char *ext);
+
+/* transition to new source. assumes cur has already been set to new file */
+static void do_transition(void *data, bool to_null)
+{
+	struct slideshow *ss = data;
+	struct slideshow_data *ssd = &ss->data;
+	bool valid = !!ss->data.files.num;
+
+	if (valid && ssd->use_cut) {
+		obs_transition_set(ss->transition, ssd->slides.cur.source);
+
+	} else if (valid && !to_null) {
+		obs_transition_start(ss->transition, OBS_TRANSITION_MODE_AUTO,
+				     ssd->tr_speed, ssd->slides.cur.source);
+
+	} else {
+		obs_transition_start(ss->transition, OBS_TRANSITION_MODE_AUTO,
+				     ssd->tr_speed, NULL);
+		set_media_state(ss, OBS_MEDIA_STATE_ENDED);
+		obs_source_media_ended(ss->source);
+	}
+
+	if (valid && !to_null) {
+		calldata_set_int(&ssd->cd, "index", ssd->slides.cur.slide_idx);
+		calldata_set_string(&ssd->cd, "path", ssd->slides.cur.path);
+
+		signal_handler_t *sh =
+			obs_source_get_signal_handler(ss->source);
+		signal_handler_signal(sh, "slide_changed", &ssd->cd);
+	}
+}
+
+/* get a source via its slide idx in one of slideshow_data's circlebufs. *
+ * only used in get_new_source().                                        */
+static inline struct source_data *circlebuf_get_source(struct circlebuf *buf,
+						       size_t slide_idx)
+{
+	struct source_data *cur;
+	size_t count = buf->size / sizeof(*cur);
+
+	for (size_t i = 0; i < count; i++) {
+		cur = circlebuf_data(buf, i * sizeof(struct source_data));
+		if (cur->slide_idx == slide_idx)
+			return cur;
+	}
+
+	return NULL;
+}
+
+static void decode_image(void *data)
+{
+	obs_weak_source_t *weak = data;
+
+	obs_source_t *source = obs_weak_source_get_source(weak);
+	if (source) {
+		image_source_preload_image(obs_obj_get_data(source));
+		obs_source_release(source);
+	}
+
+	obs_weak_source_release(weak);
+}
+
+/* creates source from a file path. only used in get_new_source(). */
+static inline obs_source_t *create_source_from_file(struct slideshow *ss,
+						    const char *file, bool now)
+{
+	obs_data_t *settings = obs_data_create();
+	obs_source_t *source;
+
+	obs_data_set_string(settings, "file", file);
+	obs_data_set_bool(settings, "unload", false);
+	obs_data_set_bool(settings, "is_slide", !now);
```

Comment:
Elsewhere constant aliases are used for interacting with `obs_data` but these 3 lines use the actual string identifiers again.